src/emailautoconfig.cpp

```diff
@@ -0,0 +1,369 @@
+                for (int j = 0; j < keys.length(); j++)
+                    values << keys.at(j).toElement().text();
+                return values;
+            }
+        }
+    }
+    return values;
+}
+
+QString EmailAutoConfig::imapServer() const
+{
+    return configValue(QStringLiteral("incomingServer"), QStringLiteral("imap"),
+                       QStringLiteral("hostname"));
+}
+
+QString EmailAutoConfig::popServer() const
+{
+    return configValue(QStringLiteral("incomingServer"), QStringLiteral("pop3"),
+                       QStringLiteral("hostname"));
+}
+
+QString EmailAutoConfig::smtpServer() const
+{
+    return configValue(QStringLiteral("outgoingServer"), QStringLiteral("smtp"),
+                       QStringLiteral("hostname"));
+}
+
+static const char* socketTypeKeys[] = {"plain", "SSL", "STARTTLS"};
+
+int EmailAutoConfig::imapPort(QMailTransport::EncryptType type) const
+{
+    const QString value
+        = configValue(QStringLiteral("incomingServer"), QStringLiteral("imap"),
+                      QStringLiteral("port"), QString::fromLatin1(socketTypeKeys[type]));
+    bool ok;
+    int port = value.toInt(&ok);
+    return ok ? port : 0;
+}
+
+int EmailAutoConfig::popPort(QMailTransport::EncryptType type) const
+{
+    const QString value
+        = configValue(QStringLiteral("incomingServer"), QStringLiteral("pop3"),
+                      QStringLiteral("port"), QString::fromLatin1(socketTypeKeys[type]));
+    bool ok;
+    int port = value.toInt(&ok);
+    return ok ? port : 0;
+}
+
+int EmailAutoConfig::smtpPort(QMailTransport::EncryptType type) const
+{
+    const QString value
+        = configValue(QStringLiteral("outgoingServer"), QStringLiteral("smtp"),
+                      QStringLiteral("port"), QString::fromLatin1(socketTypeKeys[type]));
+    bool ok;
+    int port = value.toInt(&ok);
+    return ok ? port : 0;
+}
+
+static EmailAutoConfig::AuthList toAuthList(const QStringList &values)
+{
+    EmailAutoConfig::AuthList list;
+    for (const QString &auth : values) {
+        if (auth == QStringLiteral("password-cleartext")) {
+            list << QMail::PlainMechanism;
+        } else if (auth == QStringLiteral("password-encrypted")) {
+            list << QMail::CramMd5Mechanism;
+        } else if (auth == QStringLiteral("OAuth2")) {
+            list << QMail::XOAuth2Mechanism;
+        }
+    }
+    if (list.isEmpty())
+        list << QMail::NoMechanism;
+    return list;
+}
+
+EmailAutoConfig::AuthList EmailAutoConfig::imapAuthentication(QMailTransport::EncryptType type) const
+{
+    return toAuthList(configList(QStringLiteral("incomingServer"),
+                                 QStringLiteral("imap"),
+                                 QString::fromLatin1(socketTypeKeys[type]),
+                                 QStringLiteral("authentication")));
+}
+
+EmailAutoConfig::AuthList EmailAutoConfig::popAuthentication(QMailTransport::EncryptType type) const
```

Comment:
Hmh, these two seem only used by the unit test thouh smtp has usage in emailaccount. Maybe good enough but just noting.

Reply:
Well, I agree that EmailAccount is not using this information at the moment, and may never actually use it, relying on autodetection by CAPABILITIES in the protocols (pop or imap). But I prefer to be complete in this EmailAutoConfig class, providing accessors for every bit of information.